### PR TITLE
[SPARK-47276][PYTHON][CONNECT] Introduce `spark.profile.clear` for SparkSession-based profiling

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Introduce `spark.profile.clear` for SparkSession-based profiling.

### Why are the changes needed?
A straightforward and unified interface for managing and resetting profiling results for SparkSession-based profilers.

### Does this PR introduce _any_ user-facing change?
Yes. `spark.profile.clear` is supported as shown below.

Preparation:
```py
>>> from pyspark.sql.functions import pandas_udf
>>> df = spark.range(3)
>>> @pandas_udf("long")
... def add1(x):
...   return x + 1
... 
>>> added = df.select(add1("id"))
>>> spark.conf.set("spark.sql.pyspark.udf.profiler", "perf")
>>> added.show()
+--------+                                                                      
|add1(id)|
+--------+
...
+--------+
>>> spark.profile.show()
============================================================
Profile of UDF<id=2>
============================================================
         1410 function calls (1374 primitive calls) in 0.004 seconds
...
```

Example usage:
```py
>>> spark.profile.profiler_collector._profile_results
{2: (<pstats.Stats object at 0x7ff6484d22e0>, None)}

>>> spark.profile.clear(1)  # id mismatch
>>> spark.profile.profiler_collector._profile_results
{2: (<pstats.Stats object at 0x7ff6484d22e0>, None)}

>>> spark.profile.clear(type="memory")  # type mismatch
>>> spark.profile.profiler_collector._profile_results
{2: (<pstats.Stats object at 0x7ff6484d22e0>, None)}

>>> spark.profile.clear()  # clear all
>>> spark.profile.profiler_collector._profile_results
{}
>>> spark.profile.show()
>>> 
```

### How was this patch tested?
Unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by xinrong-meng** [comment]
Failed tests are irrelevant to changes proposed in this PR. Rerun failed tests https://github.com/xinrong-meng/spark/actions/runs/8162084262.

File: python/pyspark/sql/profiler.py [hunk 1]
```diff
@@ -224,6 +224,54 @@ def dump(id: int) -> None:
             for id in sorted(code_map.keys()):
                 dump(id)
 
+    def clear_perf_profiles(self, id: Optional[int] = None) -> None:
+        """
+        Clear the perf profile results.
+
+        .. versionadded:: 4.0.0
```
Comment (HyukjinKwon) [inline python/pyspark/sql/profiler.py:231]: Is this a user-facing API? If not, we don't need this version directive

**Review by HyukjinKwon — COMMENTED** [review]
Seems fine but cc @ueshin

File: python/pyspark/sql/profiler.py (hunk 1 above)
Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py:231]: It is a user-facing API, along with `profile.show` and `profile.dump`. We will also add it to API doc.

File: python/pyspark/sql/profiler.py [hunk 2]
```diff
@@ -224,6 +224,54 @@ def dump(id: int) -> None:
...
+        """
+        ids_to_remove = [
+            result_id
+            for result_id, (perf, _, *_) in self._profile_results.items()
+            if perf is not None
+        ]
+        with self._lock:
+            if id is not None:
+                if id in ids_to_remove:
+                    self._profile_results.pop(id, None)
```
Comment (ueshin) [inline python/pyspark/sql/profiler.py:247]: Seems to be removing `memory` as well?

Comment (ueshin) [reply python/pyspark/sql/profiler.py:247]: On Jupyter:

```py
from pyspark.sql.functions import pandas_udf
df = spark.range(3)

@pandas_udf("long")
def add1(x):
    return x + 1

added = df.select(add1("id"))

spark.conf.set("spark.sql.pyspark.udf.profiler", "perf")
added.show()

spark.conf.set("spark.sql.pyspark.udf.profiler", "memory")
added.show()

spark.profile.show()
...

spark.profile.clear(type="memory")

spark.profile.show()  # should still show the perf results?
```

Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py]: Good catch! Thanks for the example. I adjusted the code and added tests.

File: python/pyspark/sql/profiler.py [hunk 3]
```diff
@@ -236,18 +236,22 @@ def clear_perf_profiles(self, id: Optional[int] = None) -> None:
...
+                    if mem is None:
+                        self._profile_results.pop(id, None)
             else:
-                for id_to_remove in ids_to_remove:
-                    self._profile_results.pop(id_to_remove, None)
+                ids_to_remove = []
+                for id, (perf, mem, *rest) in list(self._profile_results.items()):
+                    self._profile_results[id] = (None, mem, *rest)
+                    if mem is None:
+                        ids_to_remove.append(id)
```
Comment (ueshin) [inline python/pyspark/sql/profiler.py:251]: nit: Can't we pop it here?

File: python/pyspark/sql/profiler.py [hunk 4]
```diff
@@ -262,15 +266,21 @@ def clear_memory_profiles(self, id: Optional[int] = None) -> None:
...
+                    if perf is None:
+                        self._profile_results.pop(id, None)
             else:
-                for id_to_remove in ids_to_remove:
-                    self._profile_results.pop(id_to_remove, None)
+                ids_to_remove = []
+                for id, (perf, mem, *rest) in list(self._profile_results.items()):
+                    self._profile_results[id] = (perf, None, *rest)
+                    if perf is None:
+                        ids_to_remove.append(id)
```
Comment (ueshin) [inline python/pyspark/sql/profiler.py:280]: ditto.

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM, pending tests.

File: python/pyspark/sql/profiler.py (hunk 3 above)
Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py]: Good idea! Adjusted.

File: python/pyspark/sql/tests/test_session.py [hunk 5]
```diff
@@ -531,6 +531,33 @@ def test_dump_invalid_type(self):
             },
         )
 
+    def test_clear_memory_type(self):
```
Comment (zhengruifeng) [inline python/pyspark/sql/tests/test_session.py:534]: nit, it seems we don't have a parity test for `test_session`. does it make sense to move `SparkSessionProfileTests` out of `test_session` and add parity test for it?

File: python/pyspark/sql/profiler.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/sql/profiler.py:231]: Actually this is not. The `clear` in `Profile` should be a user-facing API.

**Pull request closed without merging**